src/api/v4/error-handling/carto-error.js

```diff
@@ -8,25 +24,51 @@ var UNEXPECTED_ERROR = 'unexpected error';
  *
  * @api
  */
-function CartoError (error) {
+function CartoError (error, opts) {
+  opts = opts || {};
+  var cartoError = _.extend({
+    message: UNEXPECTED_ERROR,
+    type: '',
+    origin: GENERIC_ORIGIN
+  }, error);
+
   if (_isWindshaftError(error)) {
-    this.message = error.errors[0];
-    return this;
-  }
-  if (error && error.message) {
-    this.message = error.message;
-    return this;
+    cartoError = transformWindshaftError(error, opts.layers, opts.analysis);
   }
   if (error && error.responseText) {
-    this.message = _handleAjaxResponse(error);
-    return this;
+    cartoError.message = _handleAjaxResponse(error);
+    cartoError.origin = 'ajax';
+    cartoError.type = error.statusText;
   }
 
-  this.message = UNEXPECTED_ERROR;
+  var friendlyError = retrieveFriendlyError(cartoError);
+  track(friendlyError);
+
+  friendlyError.name = 'CartoError';
+  friendlyError.stack = (new Error()).stack;
+
+  return friendlyError;
 }
 
+// Windshaft should have been parsed already 
 function _isWindshaftError (error) {
-  return error && error.errors_with_context;
+  return error && error.origin === 'windshaft';
+}
+
+function transformWindshaftError (error, layers, analysis) {
```

Comment:
use underscore for private functions so we know they are safe to nuke 💥

Reply:
In this case it's not a function of a class. It's not being exposed since it only exports a function. No need to underscore them.

Reply:
Forget about my comment. I mistook this comment for the one in the `error-list`.